[Security Solution][Alerts] Alert (+Investigation) User Assignment (#2504)

e40pud opened the pull request:
## Summary

With this PR we introduce a new Alert User Assignment feature:
- It is possible to assign a user/s to alert/s
- There is a new "Assignees" column in the alerts table which displays avatars of assigned users
- There is a bulk action to update assignees for multiple alerts
- It is possible to see and update assignees inside the alert details flyout component
- There is an "Assignees" filter button on the Alerts page which allows to filter alerts by assignees

We decided to develop this feature on a separate branch. This gives us ability to make sure that it is thoroughly tested and we did not break anything in production. Since there is a data scheme changes involved we decided that it will be a better approach. cc @yctercero 

## Testing notes

In order to test assignments you need to create a few users. Then for users to appear in user profiles dropdown menu you need to activate them by login into those account at least once.

https://github.com/elastic/kibana/assets/2700761/8eeb13f3-2d16-4fba-acdf-755024a59fc2

Main ticket https://github.com/elastic/security-team/issues/2504

## Bugfixes
- [x] https://github.com/elastic/security-team/issues/8028
- [x] https://github.com/elastic/security-team/issues/8034
- [x] https://github.com/elastic/security-team/issues/8006
- [x] https://github.com/elastic/security-team/issues/8025
- [x] https://github.com/elastic/security-team/issues/8035

## Enhancements
- [x] https://github.com/elastic/security-team/issues/8033
- [x] https://github.com/elastic/security-team/issues/8019
- [x] https://github.com/elastic/security-team/issues/8026

### Checklist

- [x] Functional changes are hidden behind a feature flag. If not hidden, the PR explains why these changes are being implemented in a long-living feature branch.
- [x] Functional changes are covered with a test plan and automated tests.
  - [x] https://github.com/elastic/kibana/issues/171306
  - [x] https://github.com/elastic/kibana/issues/171307
- [x] Stability of new and changed tests is verified using the [Flaky Test Runner](https://ci-stats.kibana.dev/trigger_flaky_test_runner).
  - [x] https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/4091
- [x] Comprehensive manual testing is done by two engineers: the PR author and one of the PR reviewers. Changes are tested in both ESS and Serverless.
- [x] Mapping changes are accompanied by a technical design document. It can be a GitHub issue or an RFC explaining the changes. The design document is shared with and approved by the appropriate teams and individual stakeholders.
   * https://github.com/elastic/security-team/issues/7647
- [x] Functional changes are communicated to the Docs team. A ticket or PR is opened in https://github.com/elastic/security-docs. The following information is included: any feature flags used, affected environments (Serverless, ESS, or both). **NOTE: as discussed we will wait until docs are ready to merge this PR**.
   * https://github.com/elastic/security-docs/issues/4226
   * https://github.com/elastic/staging-serverless-security-docs/pull/232

e40pud committed:
Merge branch 'main' into security/feature/alert-user-assignment

e40pud committed:
[Security Solution][Detections] Extend alerts schema to accommodate the list of assigned users (#7647) (#166845)

## Summary

Closes https://github.com/elastic/security-team/issues/7647

This PR extends alert's schema. We add a new field
`kibana.alert.workflow_assignee_ids` where assignees will live.

e40pud committed:
Merge branch 'main' into security/feature/alert-user-assignment

e40pud committed:
Merge branch 'main' into security/feature/alert-user-assignment

e40pud committed:
Merge branch 'main' into security/feature/alert-user-assignment

# Conflicts:
#	packages/kbn-alerts-as-data-utils/src/schemas/generated/alert_schema.ts
#	packages/kbn-alerts-as-data-utils/src/schemas/generated/security_schema.ts

e40pud committed:
Merge branch 'main' into security/feature/alert-user-assignment

e40pud committed:
Merge branch 'main' into security/feature/alert-user-assignment

e40pud committed:
Merge branch 'main' into security/feature/alert-user-assignment

e40pud committed:
Merge branch 'main' into security/feature/alert-user-assignment

e40pud committed:
[Security Solution][Detections] Add assignees UI into alerts table (#7661) (#167079)

## Summary

Closes https://github.com/elastic/security-team/issues/7661

This PR adds Alert user assignment UI within alerts table.



https://github.com/elastic/kibana/assets/2700761/df81928a-b0c7-46ba-98b3-3803774ed239

e40pud committed:
Merge branch 'main' into security/feature/alert-user-assignment

e40pud committed:
Merge branch 'main' into security/feature/alert-user-assignment

e40pud committed:
Merge branch 'main' into security/feature/alert-user-assignment

e40pud committed:
Merge branch 'main' into security/feature/alert-user-assignment

# Conflicts:
#	x-pack/test/security_solution_cypress/cypress/e2e/detection_response/detection_alerts/alert_assignees.cy.ts

e40pud committed:
Merge branch 'main' into security/feature/alert-user-assignment

e40pud committed:
Merge branch 'main' into security/feature/alert-user-assignment

e40pud committed:
Merge branch 'main' into security/feature/alert-user-assignment

e40pud committed:
Merge branch 'main' into security/feature/alert-user-assignment

e40pud committed:
Fix bug where we would apply only visible selection during the search (#169367)

## Summary

Fixes the bug where we would apply only visible user profile selections
instead of taking into account those which are not visible during the
search within component.

e40pud committed:
Merge branch 'main' into security/feature/alert-user-assignment

e40pud committed:
Fix broken tests (#169416)

## Summary

Fix broken tests introduced in
https://github.com/elastic/kibana/pull/169367

e40pud committed:
Merge branch 'main' into security/feature/alert-user-assignment

e40pud committed:
[Security Solution][Detections] Add assignees UI into alert's details flyout component (#7662) (#169508)

## Summary

Closes https://github.com/elastic/security-team/issues/7662

This PR adds Alert user assignment UI within alert's details flyout
component.


https://github.com/elastic/kibana/assets/2700761/b84299d7-5d65-4e9a-8836-807f51c0bbc7


This PR is a replacement to
https://github.com/elastic/kibana/pull/168467 since I broke that one
with wrong merges from main.

cc @PhilippeOberti

e40pud committed:
Locked Status and Assignee Controls for Alert Page (#7820) (#169235)

## Summary

Closes https://github.com/elastic/security-team/issues/7820

This PR adds "filter by assignees" component to the Alerts page.


https://github.com/elastic/kibana/assets/2700761/3f2c27bd-503f-4f0d-86ce-4e1f949db182

e40pud committed:
[Security Solution][Detections] UI and tests fixes of alert user assignments (#169534)

## Summary

A few fixes:
1. Broken test fix as a followup to
https://github.com/elastic/kibana/pull/169235
2. Make user profiles popover of a fixed size of 414px

e40pud committed:
Merge branch 'main' into security/feature/alert-user-assignment

e40pud committed:
[Security Solution] Move user profiles related hooks into a separate folder (#169645)

## Summary

These changes move user profiles hooks into a separate folder. Before it
was part of the `containers/detection_engine/alerts/`.

e40pud committed:
Merge branch 'main' into security/feature/alert-user-assignment

# Conflicts:
#	x-pack/plugins/security_solution/public/timelines/components/side_panel/event_details/flyout/index.tsx

e40pud committed:
Merge branch 'main' into security/feature/alert-user-assignment

e40pud committed:
Merge branch 'main' into security/feature/alert-user-assignment

e40pud committed:
Merge branch 'main' into security/feature/alert-user-assignment

e40pud committed:
Merge branch 'main' into security/feature/alert-user-assignment

e40pud committed:
Address UI/UX feedback (#170283)

## Summary

- UI/UX feedback
- Move assignees related UI in a common components folder

Main ticket https://github.com/elastic/security-team/issues/2504

e40pud committed:
Merge branch 'main' into security/feature/alert-user-assignment

e40pud committed:
Merge branch 'main' into security/feature/alert-user-assignment

e40pud committed:
Efficiently render user display names instead of IDs in the alerts table (#170358)

## Summary

Cherry-pick from @marshallmain branch to render cells efficiently inside
the alerts table.

Co-authored-by: Marshall Main <marshall.main@elastic.co>

e40pud committed:
Bring current user to the top of the user profiles list (#170321)

## Summary

UI enhancement which shows current user at the top of the user profiles
list.

Main ticket https://github.com/elastic/security-team/issues/2504

e40pud committed:
Move user profiles into a common folder

e40pud committed:
Merge branch 'main' into security/feature/alert-user-assignment

e40pud committed:
Efficiently render user avatars (#170372)

## Summary

Improved Assignees cell rendering within the alerts table.

e40pud committed:
Merge branch 'main' into security/feature/alert-user-assignment

e40pud committed:
Make alerts table row to fit user profile avatars by default (#170505)

## Summary

Allow alerts table to render user avatars without cutting them off.

Broken:
<img width="1590" alt="Screenshot 2023-11-03 at 12 12 05"
src="https://github.com/elastic/kibana/assets/2700761/319e6391-4e09-4350-87cc-391d18d15927">

Fixed:
<img width="1600" alt="Screenshot 2023-11-03 at 12 11 52"
src="https://github.com/elastic/kibana/assets/2700761/74258e7a-c7bb-4caf-8d90-367285213165">

Main ticket https://github.com/elastic/security-team/issues/2504

e40pud committed:
Use `useKibana` in user profiles hooks (#170523)

## Summary

Update user profiles hooks to use `useQuery`.

e40pud committed:
Merge branch 'main' into security/feature/alert-user-assignment

e40pud committed:
Update schema version from 8.11 to 8.12 (#170576)

## Summary

We pushed feature release from 8.11 to 8.12, thus need to update kibana
version in which new schema will be introduced.

elasticmachine commented:
Pinging @elastic/security-solution (Team: SecuritySolution)

e40pud commented on x-pack/plugins/security_solution/public/detections/components/alerts_table/index.tsx:
```diff
@@ -197,14 +197,11 @@ export const AlertsTableComponent: FC<DetectionEngineAlertTableProps> = ({
-  }, [isEventRenderedView]);
+  const rowHeightsOptions: EuiDataGridRowHeightsOptions = useMemo(() => {
+    return {
+      defaultHeight: 'auto',
```
@elastic/security-threat-hunting-investigations Let me know if this is the right way to increase row height to fit avatars by default.

e40pud replied:
It looks like EUI team updated grid component and now table's row does not cut off avatar anymore. Reverted my related changes.

e40pud committed:
Merge branch 'main' into security/feature/alert-user-assignment

e40pud committed:
Revert "Make alerts table row to fit user profile avatars by default (#170505)"

This reverts commit 89ec60ac0a23c91870893a06ea13ac37cea02776.

e40pud committed:
Merge branch 'main' into security/feature/alert-user-assignment

e40pud committed:
Merge branch 'main' into security/feature/alert-user-assignment

e40pud committed:
Merge branch 'main-upstream' into security/feature/alert-user-assignment-UPSTREAM

# Conflicts:
#	x-pack/plugins/security_solution/public/flyout/document_details/right/components/header_title.test.tsx
#	x-pack/plugins/security_solution/public/flyout/document_details/right/components/header_title.tsx
#	x-pack/plugins/security_solution/public/flyout/document_details/right/header.tsx

e40pud committed:
Merge branch 'main' into security/feature/alert-user-assignment

e40pud committed:
Merge branch 'main' into security/feature/alert-user-assignment

e40pud committed:
Fix broken tests

Followup to [Security Solution] Removing cleanKibana method from Cypress #170636 (https://github.com/elastic/kibana/pull/170636)

PhilippeOberti commented:
I think there is a raised condition issue somewhere, when adding or removing an assignee in the flyout, sometimes the UI doesn't reflect the change. I'm guessing this is due to the flyout refreshing before the change is actually saved somewhere?

https://github.com/elastic/kibana/assets/17276605/dc32fd52-dd6b-4c85-a852-c33e540f743c


https://github.com/elastic/kibana/assets/17276605/ca9448d5-3ecd-4837-af01-35a8ea756fb9

PhilippeOberti commented:
We should sync with the UIUX team on the display of the badge/avatar. The [mocks](https://www.figma.com/file/ubvhBGHee58diJNvSiy0GZ/%5B8.%2B%5D-%5BAlerts%5D-Expandable-Event-Flyout?type=design&node-id=4000-541651&mode=design&t=KGAXneLYJyszczw9-0) show a white border, but looking at EUI [badge](https://eui.elastic.co/#/display/badge) or [avatar](https://eui.elastic.co/#/display/avatar) components, I understand that they don't offer this flexibility.
We should then either remove the border from the mocks, or if it is required we should build our own component?

Us
![Screenshot 2023-11-07 at 12 08 10 PM](https://github.com/elastic/kibana/assets/17276605/22400f21-5ac0-4d37-8385-450c6db8d346)

Mocks
![Screenshot 2023-11-07 at 12 08 17 PM](https://github.com/elastic/kibana/assets/17276605/17723957-c98b-4843-a582-6409f706d69e)

PhilippeOberti commented:
For more than 2 assignees, the UI differs slightly from the [mocks](https://www.figma.com/file/ubvhBGHee58diJNvSiy0GZ/%5B8.%2B%5D-%5BAlerts%5D-Expandable-Event-Flyout?type=design&node-id=4000-541651&mode=design&t=KGAXneLYJyszczw9-0)

Us
![Screenshot 2023-11-07 at 12 11 19 PM](https://github.com/elastic/kibana/assets/17276605/929a4640-9480-4ee0-8845-38e9b9313a15)

Mocks
![Screenshot 2023-11-07 at 12 11 31 PM](https://github.com/elastic/kibana/assets/17276605/8daa65be-b6ec-4f56-afe5-c388ede1c919)

PhilippeOberti commented:
Wouldn't we want to use the same way to display the values between the flyout and the table? The component in the table isn't really handling small width gracefully...
![Screenshot 2023-11-07 at 12 15 02 PM](https://github.com/elastic/kibana/assets/17276605/558d421a-876c-49d7-995a-cba9839bdf73)

PhilippeOberti commented:
Just curious on the decision to have the `Assignees` dropdown to filter the table next to the `Manage rules` button. The same filtering can be done by adding a page filter (granted that the page filter currently displays an id instead of a user, but maybe that can be improved?)

https://github.com/elastic/kibana/assets/17276605/dc769006-0fdf-4fe0-9197-9d7ae0913166

PhilippeOberti commented on x-pack/plugins/security_solution/public/flyout/document_details/right/components/assignees.tsx:
```diff
@@ -0,0 +1,131 @@
+    );
+
+    return (
+      <EuiFlexGroup alignItems="center" direction="row" gutterSize="xs">
```
you will want to add `responsive={false}` here to avoid having a weird look in small screens
Normal screen size
![Screenshot 2023-11-07 at 3 01 19 PM](https://github.com/elastic/kibana/assets/17276605/b8d3e33e-d375-49bb-b8a2-49d55adf301e)

Small Screen
![Screenshot 2023-11-07 at 3 01 26 PM](https://github.com/elastic/kibana/assets/17276605/e9db5b8d-ff18-4aba-8569-f5a53ee441e8)

I have the same changes to the rest of the header in another PR, waiting to be reviewed

PhilippeOberti reviewed (commented):
I started the review but as the PR is pretty big and involved I'll get back to it tomorrow!

vitaliidm reviewed (commented):
Thanks for enormous work on this new feature.
I haven't looked at code, since it has already assignee from our team, but wanted to play  a bit with a new functionality. Few small questions around alert assignee functionality behavior


1. It seems like I can press Apply assignees to the empty list of selections. Should this button be disabled?

https://github.com/elastic/kibana/assets/92328789/f35a6fed-e6ee-4b45-9a49-ce4f1913faa9


2. Sorting seems not working for this column

https://github.com/elastic/kibana/assets/92328789/2746701c-a3e3-423d-b5db-5190a012082f


3. In comparison with alerts tags, alerts assignees do not have the third status of selection(partial selection "-"). Is this intentional?


https://github.com/elastic/kibana/assets/92328789/17aaf49a-0806-4326-9d42-41422521ce41

e40pud committed:
Merge branch 'main' into security/feature/alert-user-assignment

e40pud committed:
Merge branch 'main' into security/feature/alert-user-assignment

e40pud commented:
@PhilippeOberti thanks for the review!! I will add all answers in this comment:

> I think there is a raised condition issue somewhere, when adding or removing an assignee in the flyout, sometimes the UI doesn't reflect the change. I'm guessing this is due to the flyout refreshing before the change is actually saved somewhere?

Yes, we've seen this behaviour and did not find it a blocker for the release. I created a ticket to track this issue: https://github.com/elastic/security-team/issues/8006

> We should sync with the UIUX team on the display of the badge/avatar. The [mocks](https://www.figma.com/file/ubvhBGHee58diJNvSiy0GZ/%5B8.%2B%5D-%5BAlerts%5D-Expandable-Event-Flyout?type=design&node-id=4000-541651&mode=design&t=KGAXneLYJyszczw9-0) show a white border, but looking at EUI [badge](https://eui.elastic.co/#/display/badge) or [avatar](https://eui.elastic.co/#/display/avatar) components, I understand that they don't offer this flexibility.

We discussed avatars/badges with design team (Fernanda and Sergey) and Paul. We do not want to create a different UI/UX experience from the rest of the app. We decided to go with the current implementation and stick to standard components provided by EUI team. If we want to change the look of avatars we would need to build a sharable component used within kibana.

> Wouldn't we want to use the same way to display the values between the flyout and the table? The component in the table isn't really handling small width gracefully...

Design team decision (specifically Sergei Sivukhin) was to show badge for the 3+ assignees within the flyout and 5+ assignees within the alerts table. I think that will cover most of the cases and not sure whether we would like to dynamically adjust 5+ number based on column width.

> Just curious on the decision to have the Assignees dropdown to filter the table next to the Manage rules button. The same filtering can be done by adding a page filter (granted that the page filter currently displays an id instead of a user, but maybe that can be improved? @logeekal)

Currently, filter control button that we use on the Alerts page are tightly bound to work with dataview/field name only. I discussed with the responsible team and it is not easy to adjust that. We tried to mimic those buttons to re-use it, but it involves lots of duplication and at the end brings more issues than benefits. The decision was to add a separate filter button for assignees case next to the "Manage rules" button. In future we'd like to ask a team responsible for filter control buttons to generalise those and allow providing data source.

e40pud commented:
@vitaliidm thanks for the review!!

Here are the answers:

> It seems like I can press Apply assignees to the empty list of selections. Should this button be disabled?

I mimiced exact behaviour of the "tags" menu. I will have a look what we can do about it.

> Sorting seems not working for this column

It works to me, though the results might not be as obvious and sorting by this column might not make much of a sense. Behind each user avatar in the column there is a UID.

https://github.com/elastic/kibana/assets/2700761/52446502-93d9-4bb1-a45d-9e7bfbd994a2

> In comparison with alerts tags, alerts assignees do not have the third status of selection(partial selection "-"). Is this intentional?

Yup, that is a known issue which is not easily to address before FF. The component that we use to display user profiles list is developed outside of security solution and used in some other places (like Cases). When developed there were no requirements to support mixed states (shown as "-") thus there is non right now. Responsible team confirmed that and said it won't be an easy change. After discussion with Paul and design team, we decided that it is not a blocker. I created a separate ticket to track that issue https://github.com/elastic/security-team/issues/8008

e40pud committed:
Review feedback

e40pud commented on x-pack/plugins/security_solution/public/detections/configurations/security_solution_detections/fetch_page_context.tsx:
The intention with of this functionality to make sure that we do not fetch user profiles for each row in the alerts table. Instead we just fetch all the relevant (visible in the table) user profiles and we will pass those profiles to the column renderer.

logeekal commented:
>Just curious on the decision to have the Assignees dropdown to filter the table next to the Manage rules button. The same filtering can be done by adding a page filter (granted that the page filter currently displays an id instead of a user, but maybe that can be improved? @logeekal)

@PhilippeOberti , @e40pud already started the discussion regarding that but currently Page filters are not equipped to handle this requirement. You can read all the details [here](https://elastic.slack.com/archives/C02EDDT1M35/p1696930030428489)

logeekal commented:
>In future we'd like to ask a team responsible for filter control buttons to generalise those and allow providing data source.

@e40pud , I think Filter Controls do support custom dataview or data source.. but I think it has limit on the customization of how the options render. Here is the discussion regarding that. I am just reading through the discussion you started [here](https://elastic.slack.com/archives/C02EDDT1M35/p1696930030428489).

vitaliidm commented:
> It works to me, though the results might not be as obvious and sorting by this column might not make much of a sense. Behind each user avatar in the column there is a UID.


Wondering, is this possible to just disable sorting for that column then?


>  After discussion with Paul and design team, we decided that it is not a blocker. I created a separate ticket to track that issue https://github.com/elastic/security-team/issues/8008

Thanks for creating the issue. Together with my point 1, it creates the following behaviour: when applying empty selection of assignees, it has different effect on the alerts, affecting only those, that have assignee selected in each alert, but not selected  in few of them only

https://github.com/elastic/kibana/assets/92328789/41b0560f-676d-4ac2-9463-81ffe2dda327

e40pud commented on x-pack/plugins/security_solution/public/detections/hooks/trigger_actions_alert_table/use_assignees_actions.tsx:
done

e40pud commented:
@vitaliidm yes the last video is expected without mixed state support. In first step, you did not select anything that's why we count that as nothing has changed. On second step you deselect option which was removed then from both alerts. It is confusing because the absence of mixed states, but it is something we gonna look next.

As for the sorting, I will look if it is possible to disable it for the specific column.

e40pud committed:
Merge branch 'main' into security/feature/alert-user-assignment

e40pud committed:
Merge branch 'main' into security/feature/alert-user-assignment

e40pud committed:
Fix broken test

e40pud committed:
Merge branch 'main' into security/feature/alert-user-assignment

e40pud committed:
Merge branch 'main' into security/feature/alert-user-assignment

e40pud committed:
Merge branch 'main' into security/feature/alert-user-assignment

e40pud committed:
Merge branch 'main' into security/feature/alert-user-assignment

e40pud committed:
[Security Solution][Detections] Disable alert assignees updates for VIEWER role (#8019)

e40pud committed:
Merge branch 'main' into security/feature/alert-user-assignment

PhilippeOberti commented:
Would we want to have some cell actions to the assignee component in the expandable flyout? It should be easy to add and would be consistent with some other section of the header, like severity and status...

https://github.com/elastic/kibana/assets/17276605/d3304139-a3a1-471e-a1e3-a8a1ed4763c3

e40pud committed:
Fix broken tests

e40pud committed:
Merge branch 'main' into security/feature/alert-user-assignment

PhilippeOberti commented:
> @PhilippeOberti thanks for the review!! I will add all answers in this comment:
> 
> > I think there is a raised condition issue somewhere, when adding or removing an assignee in the flyout, sometimes the UI doesn't reflect the change. I'm guessing this is due to the flyout refreshing before the change is actually saved somewhere?
> 
> Yes, we've seen this behaviour and did not find it a blocker for the release. I created a ticket to track this issue: [elastic/security-team#8006](https://github.com/elastic/security-team/issues/8006)
> 

While doing more testing, I just had this behavior happen a few times, where both the flyout and the table were not refreshing. Only a full page refresh made things reflect the updated assignee. I would consider this a block no?

https://github.com/elastic/kibana/assets/17276605/b9d7dd98-1996-4a8d-9700-5241179612e5

(I've had this happen on adding an assignee to an alert that didn't have any, but also adding an assignee to an alert with already existing assignees)

e40pud commented:
@PhilippeOberti Thanks again for the review!

> Would we want to have some cell actions to the assignee component in the expandable flyout? It should be easy to add and would be consistent with some other section of the header, like severity and status...

We do not want to add extra functionality in this PR, only the one that was initially discussed/planned. Any extra functionality we can add as enhancement in consequent PRs.

> While doing more testing, I just had this behavior happen a few times, where both the flyout and the table were not refreshing. Only a full page refresh made things reflect the updated assignee. I would consider this a block no?

Yeah, I don't think it is a blocker. We can definitely look into this as a first thing after we merge this PR. Anyway, I will double check and confirm this with the team and Paul.

PhilippeOberti commented:
I see a question mark instead of the assignee in timeline. Is this behavior expected or am I doing something wrong?

https://github.com/elastic/kibana/assets/17276605/fa19bb78-044d-418a-9144-5ba6408288a4

PhilippeOberti commented:
Also, I understand that adding the cell actions to the flyout could be considered outside of the scope of this PR, but what about timeline?

https://github.com/elastic/kibana/assets/17276605/be1cf879-953d-4342-b1b7-71c7ae925c28

PhilippeOberti reviewed (commented):
Overall the code looks great, thanks for making this massive feature! I've left a couple of small comments.

Also, you already have a few components, interfaces and hooks that are code documented, but not all of them. I feel like - especially for the files under the `common` folder meant to be used by others - having some documentation would be useful. Could you do a run through all these files and add some?

Finally, I've been rethinking about [the issue](https://github.com/elastic/kibana/pull/170579#issuecomment-1806096620) related to the UI not updating correctly, I feel like we should be careful merging this new feature with a bug that can be potentially pretty annoying. As soon as this is merged it will be deployed within the next Serverless deployment, which means we could have users seen this pretty soon.

e40pud committed:
Merge branch 'main' into security/feature/alert-user-assignment

shahzad31 reviewed (approved):
Obs-UX-MG LGTM !!

e40pud commented:
@PhilippeOberti the `?` is indication of an unknown user which is a bug. Added ticket for that https://github.com/elastic/security-team/issues/8028

e40pud committed:
[Security Solution][Detections] Assignees shown as Unknown in Timelines (#8028)

e40pud committed:
Fix broken test

e40pud committed:
[Security Solution][Detections] Remove "Assignees" option in rule preview workflow (#8034)

e40pud committed:
Merge branch 'main' into security/feature/alert-user-assignment

e40pud committed:
Add "Remove all assignees" button for individual alert as well (#8033) (#171195)

## Summary

Addresses https://github.com/elastic/security-team/issues/8033

These changes add "Remove all assignees" button for individual alert in
alerts table.

<img width="1718" alt="Screenshot 2023-11-14 at 14 24 04"
src="https://github.com/elastic/kibana/assets/2700761/d3d1488d-5b63-4b36-9e25-cc5d75ca1310">

e40pud committed:
Merge branch 'main' into security/feature/alert-user-assignment

e40pud committed:
[Security Solution][Detections] Sometimes alerts table and details flyout do not show latest data after updating alert assignments (#171237)

## Summary

These changes fixes the issue when we show stale data after the
assignees update for the alerts in a table or details flyout.

Address:
- https://github.com/elastic/security-team/issues/8025
- https://github.com/elastic/security-team/issues/8006

cc @PhilippeOberti

yctercero commented on x-pack/plugins/security_solution/public/common/components/toolbar/bulk_actions/use_bulk_alert_assignees_items.tsx:
```diff
@@ -0,0 +1,161 @@
+  }, [onAssigneesUpdate]);
+
+  const onActionClick = useCallback<Required<BulkActionsConfig>['onClick']>(
+    async (items, isSelectAllChecked, setAlertLoading, clearSelection, refresh) => {
```
I think some of these params aren't used, like `clearSelection` and `refresh`.

e40pud replied:
done

yctercero commented on x-pack/plugins/security_solution/common/api/detection_engine/model/schemas.ts:
```diff
@@ -107,3 +110,13 @@ export const alert_tags = t.type({
 export type AlertTags = t.TypeOf<typeof alert_tags>;
+
+export const alert_assignees = t.type({
+  assignees_to_add: t.array(t.string),
```
Should we use non empty string type? So you can't send `['']`.

e40pud replied:
done

yctercero commented on x-pack/plugins/security_solution/public/detections/containers/detection_engine/alerts/api.ts:
```diff
@@ -256,3 +259,19 @@ export const getHostMetadata = async ({
+/**
+ * Fetches suggested user profiles
+ */
+export const suggestUsers = async ({
```
I think we defined this exact function somewhere else. Can we put it somewhere it can be reused?

e40pud replied:
Removed! It was a leftover

yctercero commented on x-pack/plugins/security_solution/public/detections/containers/detection_engine/alerts/types.ts:
```diff
@@ -118,3 +118,7 @@ export interface Privilege {
   has_encryption_key: boolean;
 }
+
+export interface SuggestUsersProps {
```
Can we pull from the server side types?

e40pud replied:
Removed! It was a leftover

yctercero commented on x-pack/plugins/security_solution/server/lib/detection_engine/routes/signals/helpers.ts:
```diff
@@ -20,3 +20,16 @@ export const validateAlertTagsArrays = (tags: AlertTags, ids: string[]) => {
+export const validateAlertAssigneesArrays = (assignees: AlertAssignees, ids: string[]) => {
+  const validationErrors = [];
+  if (ids.length === 0) {
+    validationErrors.push(i18n.NO_IDS_VALIDATION_ERROR);
```
I think we can have this validation in the type itself by specifying that it should be a non-empty string array.

e40pud replied:
done

yctercero commented on x-pack/plugins/triggers_actions_ui/public/application/alert_table_config_registry.ts:
```diff
@@ -15,7 +15,7 @@ import {
   private readonly objectTypes: Map<
     string,
-    AlertsTableConfigurationRegistry | AlertsTableConfigurationRegistryWithActions
+    AlertsTableConfigurationRegistry<any> | AlertsTableConfigurationRegistryWithActions
```
Any way to type this `any`?

yctercero reviewed (commented):
Thanks for this big effort and setting up time to test as well. I've left a couple of comments. Overall, I think it looks great! I just see a few things:

- It may just be my late night brain, but it seems there are some components or functions doing very similar, if not identical things that we may be able to consolidate.
- Are FTR tests in the works? I think we need to add FTRs for the new routes and include testing some RBAC to ensure someone without write privileges to the index is blocked.
- Can we add a cypress test to test adding an assignee from the alert details flyout and a test for filtering on the alerts page?
- Can you expand in the PR description what information would be helpful for your future self or future devs to know if they need to dig into this feature? What are some of the nuances around the profile endpoint right now? You had mentioned that API users do not get a profile until they log into the UI, right? When is it expected behavior to have the `id` values show up vs. the user names? 
- Can we add a task in the PR description to track the docs work? We'll need to ensure that the docs are ready to go before merging. 

I know we're working with QA already on this one to create a testing plan, but I'm wondering if it would be useful for features moving forward to include a GitHub issue with the testing plan that outlines which are automated and which are only covered manually. Could we update the PR description to mention which scenarios are NOT covered in the automated tests?

Thanks for the patience here as we use this feature to figure out some of the new processes!

e40pud committed:
Merge branch 'main' into security/feature/alert-user-assignment

e40pud commented:
Thanks for the feedback @yctercero !

> It may just be my late night brain, but it seems there are some components or functions doing very similar, if not identical things that we may be able to consolidate.

**Will look into that.**

> Are FTR tests in the works? I think we need to add FTRs for the new routes and include testing some RBAC to ensure someone without write privileges to the index is blocked.

I created two tasks to cover test plans and actual tests:
- https://github.com/elastic/kibana/issues/171306
- https://github.com/elastic/kibana/issues/171307

> Can we add a cypress test to test adding an assignee from the alert details flyout and a test for filtering on the alerts page?

For sure! I still need to cover most of the functionality with the cypress tests. 

> Can you expand in the PR description what information would be helpful for your future self or future devs to know if they need to dig into this feature? What are some of the nuances around the profile endpoint right now? You had mentioned that API users do not get a profile until they log into the UI, right? When is it expected behavior to have the id values show up vs. the user names?

Will add it to the description.

> Can we add a task in the PR description to track the docs work? We'll need to ensure that the docs are ready to go before merging.

Yup it is already there :-)

> I know we're working with QA already on this one to create a testing plan, but I'm wondering if it would be useful for features moving forward to include a GitHub issue with the testing plan that outlines which are automated and which are only covered manually. Could we update the PR description to mention which scenarios are NOT covered in the automated tests?

Yup, I created tickets to cover test plans. See my comment above.

e40pud committed:
Merge branch 'main' into security/feature/alert-user-assignment

e40pud committed:
Merge branch 'main' into security/feature/alert-user-assignment

e40pud committed:
Fix broken tests after merge conflicts

e40pud committed:
Merge branch 'main' into security/feature/alert-user-assignment

e40pud committed:
Merge branch 'main' into security/feature/alert-user-assignment

e40pud commented on x-pack/plugins/security_solution/public/common/components/user_profiles/use_get_current_user.tsx:
```diff
@@ -0,0 +1,42 @@
+  return security.userProfiles.getCurrent({ dataPath: 'avatar' });
+};
+
+export const useGetCurrentUser = () => {
```
Those are both different, even though they look similar. They use different services to fetch authorized user (via `security.authc.getCurrentUser`) vs current user profiles (via `security.userProfiles.getCurrent`). To differentiate these two hooks I can rename ours and add `Profiles` (`useGetCurrentUserProfile`) at the end to reflect the fact that it is related to `security.userProfiles` service.

e40pud replied:
Updated the name of the hook to `useGetCurrentUserProfile`.

e40pud committed:
[Security Solution][Detections] Add test coverage for Alert User Assignment (#171307)

mgiota reviewed (approved):
o11y changes LGTM!

e40pud committed:
Merge branch 'main' into security/feature/alert-user-assignment

christineweng commented on x-pack/plugins/security_solution/public/flyout/document_details/right/components/header_title.tsx:
```diff
@@ -82,6 +97,15 @@ export const HeaderTitle: FC = memo(() => {
         <EuiFlexItem grow={false}>
           <RiskScore />
         </EuiFlexItem>
+        {scopeId !== TableId.rulePreview && (
```
Just want to bring your attention to this PR that uses new expandable flyout in create rule workflow. Once it is merged you will have access to `isPreview` here

christineweng reviewed (commented):
@e40pud thank you for this feature! Some comments/questions related to flyout:

1. When i desk tested with no license applied (just default basic), I can click on assign button, then a toast appears citing license error? perhaps hide/disable the button when user does not have proper license?

![image](https://github.com/elastic/kibana/assets/18648970/9f29e47c-e7fb-46c0-8b70-28c04c45f330)

2. Is the new user assignment feature only applicable for alerts? Alert table is also used in explore pages. If you go to explore => host => events, expanding any events I can see the assignment feature there.  Maybe wrap it within `isAlert` check?

https://github.com/elastic/kibana/assets/18648970/5c359482-eee5-4d66-b5ca-a578b08cc111

e40pud committed:
Merge branch 'main' into security/feature/alert-user-assignment

e40pud committed:
Review feedback

- rename `useGetCurrentUser` into `useGetCurrentUserProfile` to avoid confusion with similar hook for fetching authenticated user
- better namings
- remove redundant code

e40pud commented on x-pack/plugins/security_solution/common/api/detection_engine/alert_assignees/set_alert_assignees/set_alert_assignees_route.mock.ts:
```diff
@@ -0,0 +1,17 @@
+  assigneesToRemove: string[] = [],
+  ids: string[] = []
+): SetAlertAssigneesRequestBody => ({
+  assignees: { assignees_to_add: assigneesToAdd, assignees_to_remove: assigneesToRemove },
```
done

e40pud commented on x-pack/plugins/security_solution/common/api/detection_engine/alert_assignees/set_alert_assignees/set_alert_assignees_route.ts:
```diff
@@ -0,0 +1,20 @@
+export const setAlertAssigneesRequestBody = t.exact(
+  t.type({
+    assignees: alert_assignees,
+    ids: alert_assignee_ids,
```
done

e40pud commented on x-pack/plugins/security_solution/common/constants.ts:
```diff
@@ -314,6 +314,10 @@ export const DETECTION_ENGINE_SIGNALS_MIGRATION_STATUS_URL =
+export const DETECTION_ENGINE_ALERT_ASSIGNEES_URL =
+  `${DETECTION_ENGINE_SIGNALS_URL}/assignees` as const;
+export const DETECTION_ENGINE_ALERT_SUGGEST_USERS_URL =
+  `${DETECTION_ENGINE_SIGNALS_URL}/suggest_users` as const;
```
done

e40pud commented on x-pack/plugins/security_solution/public/common/components/assignees/assignees_apply_panel.tsx:
```diff
@@ -0,0 +1,156 @@
+}
+
+/**
+ * The popover to allow user assignees selection for the alert
```
done

e40pud commented on x-pack/plugins/security_solution/public/common/components/toolbar/bulk_actions/translations.ts:
```diff
@@ -211,3 +211,38 @@ export const ALERT_TAGS_CONTEXT_MENU_ITEM_TOOLTIP_INFO = i18n.translate(
+export const ALERT_ASSIGNEES_CONTEXT_MENU_ITEM_TOOLTIP_INFO = i18n.translate(
+  'xpack.securitySolution.bulkActions.alertAssigneesContextMenuItemTooltip',
+  {
+    defaultMessage: 'Change alert assignees options in Kibana Advanced Settings.',
```
Removed! That was a leftover, thanks for catching!

e40pud commented on x-pack/plugins/security_solution/public/common/components/toolbar/bulk_actions/use_bulk_alert_assignees_items.tsx:
```diff
@@ -0,0 +1,161 @@
+export const useBulkAlertAssigneesItems = ({
+  onAssigneesUpdate,
+}: UseBulkAlertAssigneesItemsProps) => {
+  const { hasIndexWrite } = useAlertsPrivileges();
```
The plan was to support in both basic/essentials. Though, due to `suggest user profiles` APIs not being available for basic license, we most probably will need to increase supported license to `platinum`.

More details are here https://github.com/elastic/dev/issues/1952#issuecomment-1822327423

semd replied:
Shouldn't we check the Kibana privilege as well? To make sure the user role is not read-only. 

Also, in Serverless, does the privilege to use assignments need to be customized by role? I see the row in the roles spreadsheet, but it's empty. Could we clarify that?

semd replied:
@yctercero Oh, yes that's weird. Letting Security read-only users update alerts looks like a privilege leak 😓. Do you know why is that?
Also, I created [this issue](https://github.com/elastic/kibana/pull/170579#:~:text=I%20just%20created-,this%20issue,-.) about alerts status (open/close/ack) updates. If this is the expected behavior I will close it.

semd replied:
@yctercero Yeap! I talked with Paul about it, there's the conclusion. Thanks!
https://github.com/elastic/kibana/issues/171905#issuecomment-1834261056

e40pud commented on x-pack/plugins/security_solution/public/common/components/toolbar/bulk_actions/use_bulk_alert_assignees_items.tsx:
```diff
@@ -0,0 +1,161 @@
+
+  const onActionClick = useCallback<Required<BulkActionsConfig>['onClick']>(
+    async (items, isSelectAllChecked, setAlertLoading, clearSelection, refresh) => {
+      const ids: string[] | undefined = items.map((item) => item._id);
```
done

e40pud commented on x-pack/plugins/security_solution/public/common/components/toolbar/bulk_actions/use_bulk_alert_assignees_items.tsx:
```diff
@@ -0,0 +1,161 @@
+    onAssigneesUpdate?.();
+  }, [onAssigneesUpdate]);
+
+  const onActionClick = useCallback<Required<BulkActionsConfig>['onClick']>(
```
done

e40pud commented on x-pack/plugins/security_solution/public/common/components/toolbar/bulk_actions/use_bulk_alert_assignees_items.tsx:
```diff
@@ -0,0 +1,161 @@
+      hasIndexWrite
+        ? [
+            {
+              id: 2,
```
Yup that's only to be unique within the panel.

e40pud commented on x-pack/plugins/security_solution/public/common/components/user_profiles/translations.ts:
```diff
@@ -0,0 +1,25 @@
+);
+
+export const UNKNOWN_USER_PROFILE_NAME = i18n.translate(
+  'xpack.securitySolution.userProfiles.unknownUser.displayName',
```
done

e40pud commented on x-pack/plugins/security_solution/public/common/components/user_profiles/use_suggest_users.tsx:
```diff
@@ -0,0 +1,44 @@
+  return suggestUsers({ searchTerm });
+};
+
+export const useSuggestUsers = ({ searchTerm }: { searchTerm: string }) => {
```
Those three are using different endpoints as you mentioned and also do different things. Unfortunately, we cannot reuse one for all of the cases. We could potentially re-use `use_bulk_get_user_profiles` for fetching current user if know it's UID, but I feel like it will be unnecessary enhancement.

`use_suggest_users` works in a way that it provides the list of available users (defaults to 10 users) in the project and allows to search for more.

`use_bulk_get_user_profiles` allows us to fetch the list of known users (known UIDs) which is beneficial in case where we have all UIDs, like in case with alerts table where we display users avatars.

e40pud committed:
Remove assignees UI from the Explore page

e40pud commented:
@christineweng Thanks for the review and findings!!

> When i desk tested with no license applied (just default basic), I can click on assign button, then a toast appears citing license error? perhaps hide/disable the button when user does not have proper license?

Yes, we are thinking how to fix it properly. Most probably it will be like you suggested where we will either disable or remove menu items and add assignees buttons.

> Is the new user assignment feature only applicable for alerts? Alert table is also used in explore pages. If you go to explore => host => events, expanding any events I can see the assignment feature there. Maybe wrap it within isAlert check?

Fixed, thanks for catching this!

e40pud commented on x-pack/plugins/security_solution/public/detections/configurations/security_solution_detections/fetch_page_context.tsx:
```diff
@@ -0,0 +1,43 @@
+  'kibana.alert.workflow_user',
+];
+
+export const useFetchPageContext: PreFetchPageContext<RenderCellValueContext> = ({
```
I think the idea was to use this as a general hook for any kind of information which needs pre-fetching in alerts table. Like in our case, we need to pre-fetch all user profiles.

@marshallmain do you have any additional comments regarding it or did I cover the main idea?

e40pud committed:
Wording feedback

According to this document: https://docs.google.com/document/d/1j01MQkk0sdSvcn6iflqxlhnVCKBdpBneyLsiE4vruVg/edit

e40pud committed:
Hide assigning functionality in Basic license

e40pud committed:
Fix broken tests

e40pud committed:
Merge branch 'main' into security/feature/alert-user-assignment

e40pud committed:
Disable "Unassign alert" on query all alerts to align with other context menu items

e40pud committed:
[Security Solution][Detections] Write a test plan for Alert User Assignment (#171306) (#171337)

## Summary

Addresses https://github.com/elastic/kibana/issues/171306

Test plans for the Alert User Assignment feature.

e40pud committed:
Review feedback: Better typings in `triggers_actions_ui`

e40pud committed:
Revert "[Security Solution][Detections] Add test coverage for Alert User Assignment (#171307)"

This reverts commit 695744bc931b833461b1026769a4d167e622f79a.

e40pud committed:
Merge branch 'main' into security/feature/alert-user-assignment

e40pud committed:
Merge branch 'main' into security/feature/alert-user-assignment

XavierM commented on x-pack/plugins/triggers_actions_ui/public/application/alert_table_config_registry.ts:
```diff
@@ -28,7 +28,7 @@ export class AlertTableConfigRegistry {
    * Registers an object type to the type registry
    */
-  public register(objectType: AlertsTableConfigurationRegistry) {
+  public register<T extends object>(objectType: AlertsTableConfigurationRegistry) {
```
I think we can remove it
```suggestion
  public register(objectType: AlertsTableConfigurationRegistry) {
```

XavierM commented on x-pack/plugins/triggers_actions_ui/public/application/sections/alerts_table/alerts_table.tsx:
```diff
@@ -159,6 +159,11 @@ const AlertsTable: React.FunctionComponent<AlertsTableProps> = (props: AlertsTab
       options: props.alertsTableConfiguration.useActionsColumn,
     });
 
+  const renderCellContext = props.alertsTableConfiguration.useFetchPageContext?.({
```
should we put this around a `useMemo`?

e40pud replied:
This is already a hook which does memoization inside. I tried to `useMemo` on top of that, but React does not like that and throws an exception :-)

semd reviewed (commented):
@e40pud Shouldn't we check the Kibana privileges along with index privileges, to allow updating assignments? At least to make sure we don't show the assignment actions to users with a "read-only" role.

Also, in Serverless, does the assignments feature need to be customized by role? I see the `Alerts assignments` row in the roles spreadsheet, but it's empty. Could we clarify that?

e40pud committed:
Update x-pack/plugins/triggers_actions_ui/public/application/alert_table_config_registry.ts

Co-authored-by: Xavier Mouligneau <xavier.mouligneau@elastic.co>

e40pud commented:
Hey @semd, thanks for the review!

> @e40pud Shouldn't we check the Kibana privileges along with index privileges, to allow updating assignments? At least to make sure we don't show the assignment actions to users with a "read-only" role.

The idea was to allow alert assignments to anyone who can update alerts index. The same way we treat tags and status updates.

> Also, in Serverless, does the assignments feature need to be customized by role? I see the `Alerts assignments` row in the roles spreadsheet, but it's empty. Could we clarify that?

Same here, we want this feature to be available to everyone who can update alerts index. Can you send me a link to the document that you mentioned above please?

cc @yctercero @paulewing

e40pud committed:
Fix broken tests

e40pud committed:
Fix typing errors

e40pud committed:
Merge branch 'main' into security/feature/alert-user-assignment

e40pud committed:
Revert changes to Cases plugin

Updated by mistake here https://github.com/elastic/kibana/pull/170579/commits/38dbb5a7e35018fbb8e66ddf2dbad1a3febcc512

e40pud committed:
Revert changes to Cases plugin

Updated by mistake here https://github.com/elastic/kibana/pull/170579/commits/38dbb5a7e35018fbb8e66ddf2dbad1a3febcc512

semd commented:
@e40pud 
> The idea was to allow alert assignments to anyone who can update alerts index. The same way we treat tags and status updates. 

Should we allow write operations to users that have `read` access to Security in the Kibana features privileges? 

https://github.com/elastic/kibana/assets/17747913/03fda5b5-2498-41df-bb6e-47ce37ab62bd

role:
<img width="400" alt="security_readonly_role" src="https://github.com/elastic/kibana/assets/17747913/8fd11953-74d0-4d8f-ada6-558de361e462">

Also, index privileges are always verified implicitly when requesting an operation to ES, it is not strictly needed to check them explicitly, it makes sense only to hide the action in the UI, but ES won't allow the write operation anyway and the message will be notified to the user.

About the alert status actions, I just created [this issue](https://github.com/elastic/kibana/issues/171905).

e40pud committed:
Merge branch 'main' into security/feature/alert-user-assignment

e40pud committed:
Fix broken tests

e40pud committed:
Merge branch 'main' into security/feature/alert-user-assignment

e40pud committed:
[Security Solution][Detections] Add test coverage for Alert User Assignment (#171307) (#171930)

## Summary

Addresses https://github.com/elastic/kibana/issues/171307

This PR adds tests overage according to this test plans
https://github.com/elastic/kibana/issues/171306

e40pud committed:
Merge branch 'main' into security/feature/alert-user-assignment

e40pud commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into security/feature/alert-user-assignment

e40pud committed:
no-op commit

e40pud committed:
Merge branch 'main' into security/feature/alert-user-assignment

# Conflicts:
#	x-pack/plugins/security_solution/public/flyout/document_details/right/components/header_title.tsx

e40pud commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into security/feature/alert-user-assignment

yctercero commented on x-pack/plugins/security_solution/common/api/detection_engine/model/schemas.ts:
```diff
@@ -107,3 +110,13 @@ export const alert_tags = t.type({
 });
 
 export type AlertTags = t.TypeOf<typeof alert_tags>;
+
```
Wonder if it's best to move these to their own file like `alert_assignees`. Might help with generating the OpenAPI specs?

e40pud replied:
Added PR for this https://github.com/elastic/kibana/pull/172285

yctercero commented on x-pack/plugins/security_solution/public/common/components/user_profiles/use_get_current_user_profile.tsx:
```diff
@@ -0,0 +1,42 @@
+  return security.userProfiles.getCurrent({ dataPath: 'avatar' });
+};
+
+export const useGetCurrentUserProfile = () => {
```
Nit: can you also add a note of when to use this hook vs the other current user hooks that exist?

e40pud replied:
done

yctercero commented on x-pack/plugins/security_solution/public/detections/configurations/security_solution_detections/render_cell_value.tsx:
```diff
@@ -173,10 +174,11 @@ export const getRenderCellValueHook = ({
             scopeId={tableId}
             truncate={truncate}
             asPlainText={false}
+            context={context as RenderCellValueContext}
```
We've had bugs before when casting props. Is there any way around the cast here? Are there any values that could be missing in `context` that are needed in `RenderCellValue`?

e40pud replied:
The idea was to make `triggers_action_ui` agnostic to context format. That's why we put first `any` but then switched to `unknown`. This makes us do this casting on our side, since we create the context and know exact type.

yctercero commented on x-pack/plugins/security_solution/server/lib/detection_engine/routes/signals/set_alert_assignees_route.ts:
```diff
@@ -0,0 +1,123 @@
+import { buildRouteValidation } from '../../../../utils/build_validation/route_validation';
+import { validateAlertAssigneesArrays } from './helpers';
+
+export const setAlertAssigneesRoute = (router: SecuritySolutionPluginRouter) => {
```
We need to manually add the OpenAPI specs for the new routes.

e40pud replied:
Added PR for this https://github.com/elastic/kibana/pull/172285

yctercero reviewed (commented):
I think it's super super close! Great work with this much anticipated feature. I think three things on my list before LGTM are:

- Check basic license experience once updates are made for the tooltip discussed during sync
- Ensure OpenAPI updates are made - see `packages/kbn-openapi-generator/README.md`
- Run flakey test runner 

Other things that I think you've already created issues for to follow up on:
- Apply button is enabled even when no selection made 
  - doesn't look like any call is being made though on alerts table
  - call is still made from alert details
- Unassign alert button is enabled even when no assignment exists 
  - doesn't look like any call is being made though in table or alert details
- Possible UX follow-up - what should read only experience be?

e40pud committed:
Upselling tooltips for alert assignments feature within Basic license (#172014)

## Summary

These changes add upselling for "Alert assignments" feature within Basic
license:

<img width="619" alt="Screenshot 2023-11-27 at 19 51 36"
src="https://github.com/elastic/kibana/assets/2700761/b5d9cf36-d2c7-4387-a3f0-83dbf5c616e8">

**NOTE**: exact upselling message is still under design team review.

---------

Co-authored-by: Sergi Massaneda <sergi.massaneda@gmail.com>

e40pud committed:
Merge branch 'main' into security/feature/alert-user-assignment

# Conflicts:
#	x-pack/plugins/security_solution/public/timelines/components/side_panel/event_details/flyout/footer.tsx
#	x-pack/plugins/security_solution/public/timelines/components/side_panel/event_details/index.test.tsx

e40pud committed:
Trigger Deployment

e40pud commented:
Thanks for another round of reviews @yctercero !!

> I think it's super super close! Great work with this much anticipated feature. I think three things on my list before LGTM are:
> 
> * Check basic license experience once updates are made for the tooltip discussed during sync

**_It is merged, please check it out._**

> * Ensure OpenAPI updates are made - see `packages/kbn-openapi-generator/README.md`

**_I think we might want to push this to post FF. I most probably will be out for couple days (due to covid) and we risk not to deliver feature before FF in this case. What do you think of adding a separate task which we address during the FF?_**

> * Run flakey test runner

**_The link is in description. I run it already before but will run once again with the latest changes._**

> 
> Other things that I think you've already created issues for to follow up on:
> 
> * Apply button is enabled even when no selection made
>   
>   * doesn't look like any call is being made though on alerts table
>   * call is still made from alert details
> * Unassign alert button is enabled even when no assignment exists
>   
>   * doesn't look like any call is being made though in table or alert details
> * Possible UX follow-up - what should read only experience be?

**_I will double check when I'm back that all those issues above either exist or I create a ticket for them otherwise._**

e40pud committed:
Trigger Deployment

PhilippeOberti commented on x-pack/plugins/security_solution/docs/testing/test_plans/detection_response/alerts/user_assignment.md:
❤️

PhilippeOberti reviewed (approved):
Smoke tested and everything LGTM now, all the small bugs I had found earlier have been fixed!
Amazing job on this, the code is well tested, documented and easy to navigate through!

If I allowed myself one comment, I would try in the future to avoid having such involved PRs. It has made the review a bit challenging for me. Spanning review over so many weeks, and with the sheer amount of code modified and introduced, it's been difficult to remember what I tested, what was raised by me and/or others and keep track of everything.

e40pud committed:
Merge branch 'main' into security/feature/alert-user-assignment

e40pud commented on x-pack/plugins/security_solution/common/api/detection_engine/model/schemas.ts:
```diff
@@ -42,6 +42,9 @@ export const signal_status_query = t.object;
 export const alert_tag_ids = t.array(t.string);
 export type AlertTagIds = t.TypeOf<typeof alert_tag_ids>;
 
+export const alert_ids = NonEmptyArray(NonEmptyString);
```
It is not really a `UUID` formatted string. Not sure what format is used exactly, but it just looks differently and everywhere `string` used as a type for it. Here is the example of the user id `u_79HkWkwmnBH5gqFKwoxggWPjEBOur1zLPXQPEl1VBW0_0` (taken from [here](https://www.elastic.co/guide/en/elasticsearch/reference/current/security-api-activate-user-profile.html))

yctercero replied:
Ah, ok - that makes sense then. Thanks!

e40pud committed:
Add `useGetCurrentUserProfile` hook description

e40pud committed:
Merge branch 'main' into security/feature/alert-user-assignment

e40pud committed:
Trigger Build

e40pud committed:
Merge branch 'main' into security/feature/alert-user-assignment

marshallmain reviewed (commented):
Started reviewing today but didn't get far, will finish tomorrow

e40pud committed:
Review feedback

e40pud committed:
Merge branch 'main' into security/feature/alert-user-assignment

# Conflicts:
#	x-pack/test/security_solution_cypress/cypress/tasks/navigation.ts

e40pud commented on x-pack/plugins/security_solution/docs/testing/test_plans/detection_response/alerts/user_assignment.md:
```diff
@@ -0,0 +1,217 @@
+
+- The feature is **NOT** available under the Basic license
+- Assignees are stored as an array of users IDs in alert's `kibana.alert.workflow_assignee_ids` field
+- The feature is available under the Basic license
```
done

e40pud commented on x-pack/plugins/security_solution/common/api/detection_engine/alert_assignees/set_alert_assignees/set_alert_assignees_route.ts:
```diff
@@ -0,0 +1,20 @@
+
+import { alert_ids, alert_assignees } from '../../model';
+
+export const setAlertAssigneesRequestBody = t.exact(
```
Here is the PR for that https://github.com/elastic/kibana/pull/172285

e40pud replied:
done

e40pud committed:
Merge branch 'main' into security/feature/alert-user-assignment

marshallmain commented on x-pack/plugins/security_solution/public/common/components/filter_group/filter_by_assignees.tsx:
```diff
@@ -0,0 +1,93 @@
+    const [isPopoverOpen, setIsPopoverOpen] = useState(false);
+    const togglePopover = useCallback(() => setIsPopoverOpen((value) => !value), []);
+
+    const [selectedAssignees, setSelectedAssignees] =
+      useState<AssigneesIdsSelection[]>(assignedUserIds);
```
Since `selectedAssignees` is only needed for the count of selectedAssignees, I think it would be easier to understand if this state held only the count. As it stands right now, it's hard to tell quickly which component truly owns the `selectedAssignee` state since `FilterByAssignessPopover`, `DetectionEnginePageComponent`, and `AssigneesApplyPanel` all keep a copy of the selection state.

Alternatively, this state could be passed in from the parent component rather than duplicating it here.

e40pud replied:
As discussed, I will follow up on this one in a separate PR. Here is the [ticket](https://github.com/elastic/security-team/issues/8164).

marshallmain commented on x-pack/plugins/security_solution/public/common/components/assignees/assignees_apply_panel.tsx:
```diff
@@ -0,0 +1,156 @@
+          clearButtonLabel={i18n.ASSIGNEES_CLEAR_FILTERS}
+          nullOptionLabel={i18n.ASSIGNEES_NO_ASSIGNEES}
+        />
+        {onAssigneesApply && (
+          <EuiButton
+            data-test-subj={ASSIGNEES_APPLY_BUTTON_TEST_ID}
+            fullWidth
+            size="s"
+            onClick={handleApplyButtonClick}
+            isDisabled={isLoading}
+          >
+            {i18n.ASSIGNEES_APPLY_BUTTON}
+          </EuiButton>
+        )}
```
I think this component would be clearer if the "Apply" button logic is extracted out. As I understand it, the `selectedAssignees` state is really only needed when `onAssigneesApply` is provided - otherwise, the `onChange` handler updates the selected profiles in some parent component and that parent component holds the real source of truth for selected assignee state. But, since `onAssigneesApply` is optional, it takes some extra effort to make that connection.

Perhaps this component could have the Apply button logic removed, and a new wrapper component could combine this component with the apply button. The wrapper would have require `onAssigneesApply`, and this component would then require `onSelectionChange`, instead of both being optional on this component. The wrapper component would maintain the selected assignee state (for the applying assignees to alerts use case), whereas for the "filtering by assignee" use case the `DetectionEnginePageComponent` would still maintain the selected assignee state as it does now. This way the source of truth for the selection state is always in the parent of this component. The prop types would also better communicate the intended usage of this component - right now it seems that either `on AssigneesApply` or `onSelectionChange` must be provided, but they're both optional. If we have a wrapper that requires `onAssigneesApply` and this component requires `onSelectionChange` then it's clearer how they're supposed to be used.

marshallmain reviewed (commented):
Looks pretty good overall - I think the assignees panel would benefit from some refactoring to make the data model and intended usage clearer.

yctercero reviewed (approved):
Thank you for this huge effort, including a test plan, testing party and the automated tests. 

Once https://github.com/elastic/kibana/pull/172285 is merged to add in the openAPI specs, I think it's good to go! Please ensure that one is merged in before merging this one. It looks like maybe there's some further component refactoring that is suggested. As long as @marshallmain is ok with it, I think the refactor can be followed up on. 

Given the timezone differences, I'm going ahead and LGTM-ing.

e40pud committed:
Specifying the schemas for new APIs with OpenAPI (#172285)

## Summary

With these changes we specify the schemas for new alert assignments APIs
with OpenAPI.

cc @yctercero @marshallmain

e40pud committed:
Merge branch 'main' into security/feature/alert-user-assignment

# Conflicts:
#	x-pack/plugins/security_solution/server/routes/index.ts

marshallmain reviewed (approved):
LGTM, component refactoring improvements to be done in a follow up
